scripts/compareAPIs.js

```diff
@@ -2,360 +2,545 @@
 let chalk = require('chalk');
 let yargs = require('yargs');
+let Diff = require('diff');
 
 let argv = yargs
   .option('verbose', {alias: 'v', type: 'boolean'})
   .option('organizedBy', {choices: ['type', 'change']})
   .option('rawNames', {type: 'boolean'})
+  .option('package', {type: 'string'})
+  .option('interface', {type: 'string'})
+  .option('isCI', {type: 'boolean'})
+  .option('base-api-dir', {type: 'string'})
+  .option('branch-api-dir', {type: 'string'})
   .argv;
 
+let allChanged = new Set();
+
+// {'useSliderState' => [ 'SliderStateOptions', 'SliderState' ], ... }
+let dependantOnLinks = new Map();
+let currentlyProcessing = '';
+let depth = 0;
+
 compare().catch(err => {
   console.error(err.stack);
   process.exit(1);
 });
 
+
 /**
- * This takes the json files generated by the buildBranchAPI and buildPublishedAPI and diffs each of the corresponding
- * json files. It outputs a JSON string that tells if each property is an addition, removal, or addition to the API.
- * From this, we can determine if we've made a breaking change or introduced an API we meant to be private.
- * We can high level some of this information in a series of summary messages that are color coded at the tail of the run.
+ * This takes the json files generated by the buildBranchAPI and buildPublishedAPI and
+ * reconstructs our interfaces and create a graph of all dependencies between interfaces.
+ * From there, we diff the reconstructed interfaces and track all that have changes.
+ * We use the graph to communicate what interfaces have changed as a result of a
+ * dependency changing.
+ * We build up strings of the diffs and make them easy to read in both a github comment
+ * as well as the local console.
  */
 async function compare() {
-  let branchDir = path.join(__dirname, '..', 'dist', 'branch-api');
-  let publishedDir = path.join(__dirname, '..', 'dist', 'published-api');
+  let branchDir = argv['branch-api-dir'] || path.join(__dirname, '..', 'dist', 'branch-api');
+  let publishedDir = argv['base-api-dir'] || path.join(__dirname, '..', 'dist', 'base-api');
   if (!(fs.existsSync(branchDir) && fs.existsSync(publishedDir))) {
-    console.log(chalk.redBright(`you must have both a branchDir ${branchDir} and publishedDir ${publishedDir}`));
+    console.log(chalk.redBright(`you must have both a branchDir ${branchDir} and baseDir ${publishedDir}`));
     return;
   }
-  let summaryMessages = [];
 
   let branchAPIs = fg.sync(`${branchDir}/**/api.json`);
   let publishedAPIs = fg.sync(`${publishedDir}/**/api.json`);
   let pairs = [];
-  // we only care about changes to already published APIs, so find all matching pairs based on what's been published
+
+  // find all matching pairs based on what's been published
   for (let pubApi of publishedAPIs) {
     let pubApiPath = pubApi.split(path.sep);
     let sharedPath = path.join(...pubApiPath.slice(pubApiPath.length - 4));
-    let matchingBranchFile;
+    let found = false;
```

Comment:
foundAPI or apiChildExists?

Reply:
file found, it's a pretty tight loop, i think it's mostly self evident